src/neo/SmartContract/Native/NativeContract.cs

```diff
@@ -114,19 +110,24 @@ public static NativeContract GetContract(string name)
 
         internal bool Invoke(ApplicationEngine engine)
         {
-            if (!engine.CurrentScriptHash.Equals(Hash))
-                return false;
-            string operation = engine.CurrentContext.EvaluationStack.Pop().GetString();
-            Array args = (Array)engine.CurrentContext.EvaluationStack.Pop();
-            if (!methods.TryGetValue(operation, out ContractMethodMetadata method))
-                return false;
+            if (!engine.CurrentScriptHash.Equals(Hash)) return false;
+            if (!engine.TryPop(out string operation)) return false;
+            if (!engine.TryPop(out Array args)) return false;
+            if (!methods.TryGetValue(operation, out ContractMethodMetadata method)) return false;
             ExecutionContextState state = engine.CurrentContext.GetState<ExecutionContextState>();
-            if (!state.CallFlags.HasFlag(method.RequiredCallFlags))
-                return false;
-            if (!engine.AddGas(method.Price))
-                return false;
-            StackItem result = method.Delegate(engine, args);
-            engine.CurrentContext.EvaluationStack.Push(result);
+            if (!state.CallFlags.HasFlag(method.RequiredCallFlags)) return false;
+            if (!engine.AddGas(method.Price)) return false;
+            List<object> parameters = new List<object>();
```

Comment:
we can use an array because we know the size and it's faster than a list

Reply:
But we also need to consider `NeedApplicationEngine` and `NeedSnapshot`.